working/2364 - primary constructors/feature-specification.md

```diff
@@ -439,15 +515,15 @@ class B extends A { // OK.
 class B(int a) extends A(a); // Could be supported, but isn't!
 ```
 
-There are several reasons why this is not supported. First, primary constructors
-should be small and easy to read. Next, it is not obvious how the
-superconstructor arguments would fit into a mixin application (e.g., when the
-superclass is `A with M1, M2`), or how readable it would be if the
-superconstructor is named (`class B(int a) extends A.name(a);`). For instance,
-would it be obvious to all readers that the superclass is `A` and not `A.name`,
-and that all other constructors than the primary constructor will ignore the
-implied superinitialization `super.name(a)` and do their own thing (which might
-be implicit).
+There are several reasons why this is not supported. First, primary
+constructors should be small and easy to read. Next, it is not obvious how
```

Comment:
Is "should be small and easy to read" not an equally compelling argument against supporting assertions in primary constructors?

I find it pretty surprising that we'd be OK with supporting asserts but not other initializers. Seems like a fairly arbitrary limitation.

Reply:
We have concrete evidence that developers may need assertions, that was an important point for me. The assertions would be able to specify constraints on the parameters using unsurprising scope rules, there's nothing special to remember in order to use this feature.

Supporting a general initializer list is a significantly bigger step than just supporting assertions: Surely we'd then want to support the introduction of additional instance variables by means of initializer list entries, and this would probably mean that `final int x = e` will introduce a final instance variable of type `int` whereas `this.x = e` will not, and `int this.x = e` is not supported, nor is `final this.x = e`. This implies that developers can _not_ just rely on their prior experience with initializer lists, they'll need to get used to a new syntax and a new semantics of initializer lists, just for primary constructors.

Next, we'd probably support `super.someName(...)` as well (if we want to be able to say that "primary constructors can have initializer lists").

Primary constructors is a conciseness feature and I just don't think it's a good idea to keep piling up additional features, and paying for them in terms of more verbose syntax and more complex semantics. If you want something more general then write a regular constructor.

(OK, the following could actually be a useful technique: Write a primary constructor just to get started, then ask the IDE to transform it into a normal constructor, then add a body and other stuff as needed. This would allow developers to avoid _writing_ the names twice, even in the case where it is known up front that it will be a regular constructor.)

Reply:
> We have concrete evidence that developers may need assertions, that was an important point for me.

I mean, we also have concrete evidence that developers may need initializer lists and super constructor calls... :)

You can always just not use a primary constructor if you need assertions, right?

> Supporting a general initializer list is a significantly bigger step than just supporting assertions: Surely we'd then want to support the introduction of additional instance variables by means of initializer list entries,

For what it's worth, I definitely would *not* expect that to work. I would only expect the initializer list to, well, initialize fields that are already declared. I don't know if my expectation would be widely shared.

> and this would probably mean that `final int x = e` will introduce a final instance variable of type `int` whereas `this.x = e` will not, and `int this.x = e` is not supported, nor is `final this.x = e`. This implies that developers can _not_ just rely on their prior experience with initializer lists, they'll need to get used to a new syntax and a new semantics of initializer lists, just for primary constructors.

I think we could sidestep all of this by saying that, no, the initializer list only initializes. The way I would think about it is that there are two language features:

1. Function (or method, or constructor) parameters.
2. Field declarations.

Both of those bind new variables. The magic of the primary constructor feature is that it lets you combine those. But the two things it combines are always *variable declarations*. Allowing field declarations inside the initializer list would be a different thing because there you aren't already declaring something.

Of course, allowing an initializer list in a primary constructor *would* imply that users will want non-field primary constructor parameters, which is another design question we've talked about.

> Next, we'd probably support `super.someName(...)` as well (if we want to be able to say that "primary constructors can have initializer lists").

I suspect we'd just put this after the superclass name in the `extends` clause like other languages do, though that has some design challenges.

> Primary constructors is a conciseness feature and I just don't think it's a good idea to keep piling up additional features, and paying for them in terms of more verbose syntax and more complex semantics. If you want something more general then write a regular constructor.

Agreed! So let's not pile on assertions in them. :)

> (OK, the following could actually be a useful technique: Write a primary constructor just to get started, then ask the IDE to transform it into a normal constructor, then add a body and other stuff as needed. This would allow developers to avoid _writing_ the names twice, even in the case where it is known up front that it will be a regular constructor.)

Yeah, I think that would definitely be useful.